java/src/main/java/com/google/oak/transport/GrpcStreamingTransport.java

```diff
@@ -0,0 +1,138 @@
+   * and response types.
+   */
+  public GrpcStreamingTransport(
+      Function<StreamObserver<ResponseWrapper>, StreamObserver<RequestWrapper>> stream) {
+    // We expect to receive a single response message for each request.
+    this.messageQueue = new ArrayBlockingQueue<>(1);
+    StreamObserver<ResponseWrapper> responseObserver = new StreamObserver<ResponseWrapper>() {
+      @Override
+      public void onNext(ResponseWrapper response) {
+        try {
+          messageQueue.put(response);
+        } catch (Exception e) {
+          if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          logger.log(Level.WARNING, "Couldn't send server response to the message queue: " + e);
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        Status status = Status.fromThrowable(t);
+        logger.log(Level.WARNING, "Couldn't receive response: " + status);
+      }
+
+      @Override
+      public void onCompleted() {
+        logger.log(Level.FINE, "response message queue completed");
+      }
+    };
+    this.requestObserver = stream.apply(responseObserver);
+  }
+
+  /**
+   * Returns evidence about the trustworthiness of a remote server.
+   *
+   * @return {@code AttestationBundle} wrapped in a {@code Result}
+   */
+  public Result<AttestationBundle, Exception> getEvidence() {
+    try {
+      RequestWrapper requestWrapper = RequestWrapper.newBuilder()
+                                          .setGetPublicKeyRequest(GetPublicKeyRequest.newBuilder())
+                                          .build();
+      logger.log(Level.INFO, "sending get evidence request: " + requestWrapper);
+      this.requestObserver.onNext(requestWrapper);
+
+      ResponseWrapper responseWrapper = this.messageQueue.poll(10, SECONDS);
+      logger.log(Level.INFO, "received get evidence response: " + responseWrapper);
+      GetPublicKeyResponse response = responseWrapper.getGetPublicKeyResponse();
+
+      return Result.success(response.getAttestationBundle());
+    } catch (InterruptedException e) {
+      return Result.error(e);
+    }
+  }
+
+  /**
+   * Sends a request to the enclave and returns a response.
+   *
+   * @param requestBytes a serialized {@code oak.crypto.EncryptedRequest} wrapped in a {@code
+   *     Result}
+   * @return a serialized {@code oak.crypto.EncryptedResponse} wrapped in a {@code Result}
+   */
+  public Result<byte[], Exception> invoke(byte[] requestBytes) {
+    try {
+      RequestWrapper requestWrapper =
+          RequestWrapper.newBuilder()
+              .setInvokeRequest(
+                  InvokeRequest.newBuilder().setEncryptedBody(ByteString.copyFrom(requestBytes)))
+              .build();
+      logger.log(Level.INFO, "sending invoke request: " + requestWrapper);
+      this.requestObserver.onNext(requestWrapper);
+
+      ResponseWrapper responseWrapper = this.messageQueue.poll(10, SECONDS);
```

Comment:
Consider retrying, or using `poll()`. 
Or if this time-limited poll is the best option, make `10` a class-level static const and use it both here and in the `getEvidence` method.

Reply:
Does Java have some primitives for retrying?
Or it'll be just a for loop? (aso each try should probably be time limited)

Reply:
No primitives, but there are libraries you can use. In the past I've used [guava-retrying](http://rholder.github.io/guava-retrying/javadoc/1.0.5/com/github/rholder/retry/Retryer.html).

Alternatively, perhaps the code can be refactored to use Futures. Feel free to leave this for a future PR.

Reply:
Moved to a constant and added a TODO for retries